File: easy/3903_smallest_stable_index_i/solution.cpp
#include <vector>

class Solution {
public:
  int firstStableIndex(std::vector<int> &nums, int k) {
    const int numsLen = nums.size();

    std::vector<int> prefixMaxes(numsLen, nums[0]);
    std::vector<int> prefixMins(numsLen, nums[numsLen - 1]);

    for (int i = 1; i < numsLen; ++i)
      prefixMaxes[i] = std::max(prefixMaxes[i - 1], nums[i]);

    for (int i = numsLen - 2; i > -1; --i)
      prefixMins[i] = std::min(prefixMins[i + 1], nums[i]);

    for (int i = 0; i < numsLen; ++i) {
      if (prefixMaxes[i] - prefixMins[i] <= k)
        return i;
    }

    return -1;
  }
};

File: easy/3903_smallest_stable_index_i/Solution.java
class Solution {
  public int firstStableIndex(int[] nums, int k) {
    final int[] prefixMaxes = new int[nums.length];
    final int[] prefixMins = new int[nums.length];

    prefixMaxes[0] = nums[0];
    for (int i = 1; i < nums.length; ++i) {
      prefixMaxes[i] = Math.max(prefixMaxes[i - 1], nums[i]);
    }

    prefixMins[nums.length - 1] = nums[nums.length - 1];
    for (int i = nums.length - 2; i > -1; --i) {
      prefixMins[i] = Math.min(prefixMins[i + 1], nums[i]);
    }

    for (int i = 0; i < nums.length; ++i) {
      if (prefixMaxes[i] - prefixMins[i] <= k) {
        return i;
      }
    }

    return -1;
  }
}
